File: CLI Engine/internal/crypto/crypto.go
package crypto

import (
	"context"
	"crypto/aes"
	"crypto/cipher"
	"encoding/base64"
	"encoding/hex"
	"fmt"
	"os"
	"strings"

	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
	"github.com/Azure/azure-sdk-for-go/sdk/security/keyvault/azsecrets"
)

// DecryptSecret decrypts a stored secret using the active provider.
// Switch providers via SECRET_PROVIDER env var:
//
//	aes256   — local AES-256-GCM (development, default)
//	keyvault — Azure Key Vault (production)
func DecryptSecret(stored string) (string, error) {
	provider := os.Getenv("SECRET_PROVIDER")
	if provider == "keyvault" {
		return kvDecrypt(stored)
	}
	return aesDecrypt(stored)
}

// ─── AES-256-GCM (Option 1) ──────────────────────────────────────────────────

func aesDecrypt(stored string) (string, error) {
	keyB64 := os.Getenv("ENCRYPTION_KEY")
	if keyB64 == "" {
		return "", fmt.Errorf("ENCRYPTION_KEY env var not set")
	}

	key, err := base64.StdEncoding.DecodeString(keyB64)
	if err != nil {
		return "", fmt.Errorf("decoding ENCRYPTION_KEY: %w", err)
	}
	if len(key) != 32 {
		return "", fmt.Errorf("ENCRYPTION_KEY must be 32 bytes, got %d", len(key))
	}

	parts := strings.Split(stored, ":")
	if len(parts) != 3 {
		return "", fmt.Errorf("invalid AES encrypted format, expected iv:authTag:ciphertext")
	}

	iv, err := hex.DecodeString(parts[0])
	if err != nil {
		return "", fmt.Errorf("decoding iv: %w", err)
	}
	authTag, err := hex.DecodeString(parts[1])
	if err != nil {
		return "", fmt.Errorf("decoding authTag: %w", err)
	}
	enc, err := hex.DecodeString(parts[2])
	if err != nil {
		return "", fmt.Errorf("decoding ciphertext: %w", err)
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return "", fmt.Errorf("creating AES cipher: %w", err)
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", fmt.Errorf("creating GCM: %w", err)
	}

	// Go's GCM Open expects ciphertext with authTag appended
	combined := append(enc, authTag...)
	plaintext, err := gcm.Open(nil, iv, combined, nil)
	if err != nil {
		return "", fmt.Errorf("decrypting secret: %w", err)
	}
	return string(plaintext), nil
}

// ─── Azure Key Vault (Option 2) ──────────────────────────────────────────────

func kvDecrypt(secretName string) (string, error) {
	vaultURL := os.Getenv("KEYVAULT_URL")
	if vaultURL == "" {
		return "", fmt.Errorf("KEYVAULT_URL env var not set")
	}

	cred, err := azidentity.NewDefaultAzureCredential(nil)
	if err != nil {
		return "", fmt.Errorf("keyvault auth: %w", err)
	}

	client, err := azsecrets.NewClient(vaultURL, cred, nil)
	if err != nil {
		return "", fmt.Errorf("creating keyvault client: %w", err)
	}

	resp, err := client.GetSecret(context.Background(), secretName, "", nil)
	if err != nil {
		return "", fmt.Errorf("fetching secret '%s' from keyvault: %w", secretName, err)
	}
	if resp.Value == nil {
		return "", fmt.Errorf("keyvault secret '%s' has no value", secretName)
	}
	return *resp.Value, nil
}

File: dashboard/app/api/utils/crypto.ts
import crypto from 'crypto'
import { SecretClient } from '@azure/keyvault-secrets'
import { ClientSecretCredential } from '@azure/identity'

// Switch providers via SECRET_PROVIDER env var:
//   aes256   — local AES-256-GCM (development, default)
//   keyvault — Azure Key Vault (production)
const PROVIDER = process.env.SECRET_PROVIDER || 'aes256'

// ─── AES-256-GCM (Option 1) ──────────────────────────────────────────────────

function getAESKey(): Buffer {
  const key = process.env.ENCRYPTION_KEY
  if (!key) throw new Error('ENCRYPTION_KEY env var not set')
  const buf = Buffer.from(key, 'base64')
  if (buf.length !== 32) throw new Error('ENCRYPTION_KEY must be 32 bytes (base64-encoded)')
  return buf
}

function aesEncrypt(plaintext: string): string {
  const key = getAESKey()
  const iv = crypto.randomBytes(12)
  const cipher = crypto.createCipheriv('aes-256-gcm', key, iv)
  const encrypted = Buffer.concat([cipher.update(plaintext, 'utf8'), cipher.final()])
  const authTag = cipher.getAuthTag()
  return `${iv.toString('hex')}:${authTag.toString('hex')}:${encrypted.toString('hex')}`
}

function aesDecrypt(stored: string): string {
  const key = getAESKey()
  const parts = stored.split(':')
  if (parts.length !== 3) throw new Error('Invalid AES encrypted format')
  const iv = Buffer.from(parts[0], 'hex')
  const authTag = Buffer.from(parts[1], 'hex')
  const enc = Buffer.from(parts[2], 'hex')
  const decipher = crypto.createDecipheriv('aes-256-gcm', key, iv)
  decipher.setAuthTag(authTag)
  return decipher.update(enc).toString('utf8') + decipher.final('utf8')
}

// ─── Azure Key Vault (Option 2) ──────────────────────────────────────────────

function getKeyVaultClient(): SecretClient {
  const vaultUrl = process.env.KEYVAULT_URL
  if (!vaultUrl) throw new Error('KEYVAULT_URL env var not set')
  const credential = new ClientSecretCredential(
    process.env.AZURE_TENANT_ID!,
    process.env.AZURE_CLIENT_ID!,
    process.env.AZURE_CLIENT_SECRET!
  )
  return new SecretClient(vaultUrl, credential)
}

async function kvEncrypt(plaintext: string): Promise<string> {
  const client = getKeyVaultClient()
  // Use a unique name per secret — timestamp + random suffix
  const secretName = `sub-secret-${Date.now()}-${crypto.randomBytes(4).toString('hex')}`
  await client.setSecret(secretName, plaintext)
  return secretName
}

async function kvDecrypt(secretName: string): Promise<string> {
  const client = getKeyVaultClient()
  const secret = await client.getSecret(secretName)
  if (!secret.value) throw new Error(`Key Vault secret '${secretName}' has no value`)
  return secret.value
}

// ─── Public API ──────────────────────────────────────────────────────────────

export async function encryptSecret(plaintext: string): Promise<string> {
  if (PROVIDER === 'keyvault') return kvEncrypt(plaintext)
  return aesEncrypt(plaintext)
}

export async function decryptSecret(stored: string): Promise<string> {
  if (PROVIDER === 'keyvault') return kvDecrypt(stored)
  return aesDecrypt(stored)
}
